src/test/java/com/feedzai/commons/sql/abstraction/engine/impl/abs/TestRouter.java

```diff
@@ -0,0 +1,172 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.feedzai.commons.sql.abstraction.engine.impl.abs;
+
+import com.feedzai.commons.sql.abstraction.engine.DatabaseEngine;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.ServerSocket;
+import java.net.Socket;
+import java.util.Properties;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+
+import static com.feedzai.commons.sql.abstraction.engine.configuration.PdbProperties.JDBC;
+
+/**
+ * A class that acts as a router, to forward data sent between a {@link DatabaseEngine} and the DB.
+ * <p>
+ * When running tests, this offers the possibility of interrupting connections without closing them.
+ *
+ * @author José Fidalgo (jose.fidalgo@feedzai.com)
+ */
+public class TestRouter implements Closeable {
+
+    /**
+     * Logger for this class.
+     */
+    private static final Logger logger = LoggerFactory.getLogger(TestRouter.class);
+
+    /**
+     * The {@link ServerSocket} that will accept connections from PDB engines to this {@link TestRouter}.
+     */
+    private final ServerSocket serverSocket = new ServerSocket(0);
+
+    /**
+     * The {@link Socket} to connect this {@link TestRouter} to the database.
+     */
+    private final Socket socketRouterToDb = new Socket();
+
+    /**
+     * An {@link ExecutorService} to handle asynchronous tasks.
+     */
+    private final ExecutorService executor;
+
+    /**
+     * The database port.
+     */
+    private final int dbPort;
+
+    /**
+     * Whether this {@link TestRouter} is running.
+     */
+    private volatile boolean isRunning = true;
+
+    /**
+     * Constructor for this class.
+     *
+     * @param executor An {@link ExecutorService} to run tasks asynchronously.
+     * @param dbPort   The real DB server port.
+     * @throws IOException if some problem occurs creating a new {@link ServerSocket}.
+     */
+    TestRouter(final ExecutorService executor, final int dbPort) throws IOException {
+        this.executor = executor;
+        this.dbPort = dbPort;
+    }
+
+    /**
+     * Gets new DB properties with the JDBC URL patched to replace the default DB port with a new one.
+     *
+     * @param dbProperties The original DB properties.
+     * @param defaultPort  The default DB port.
+     * @param newPort      The port to use as replacement in the JDBC URL for the default DB port.
+     * @return The patched DB properties.
+     */
+    public static Properties getPatchedDbProperties(final Properties dbProperties, final int defaultPort, final int newPort) {
```

Comment:
This could be package-private.

Reply:
It could, but so could all the other public methods, or even the class itself.
Why are you picking on this specific method? :smile: 
Thing is, this TestRouter class used to be an inner class of TimeoutsTest and I had to extract it to make it available to AbstractEngineSchemaTest. This is just a test class, I don't want to restrict it like that and then require changes if some class in other package also needs to use it.

Reply:
IMO method visibility should follow the same rules as security permissions, just enough for it to work. But this is a test class so I'm ok for it to be public.

(PS. I forgot to leave the same comment in the other methods :stuck_out_tongue: )